app/src/main/java/nu/yona/app/ui/LaunchActivity.java

```diff
@@ -117,40 +118,86 @@ else if (!sharedUserPreferences.getBoolean(PreferenceConstant.STEP_PASSCODE, fal
 		}
 		else if (!TextUtils.isEmpty(sharedUserPreferences.getString(PreferenceConstant.YONA_PASSCODE, "")))
 		{
-			startNewActivity(bundle, YonaActivity.class);
+			checkForValidStoredUser();
 		}
 	}
 
 
-	private void setListeners()
+	private void checkForValidStoredUser()
 	{
-		findViewById(R.id.join).setOnClickListener(new View.OnClickListener()
+		try
 		{
-			@Override
-			public void onClick(View v)
+			URL environmentURL = new URL(YonaApplication.getEventChangeManager().getDataState().getServerUrl());
+			URL storeUserURL = new URL(YonaApplication.getEventChangeManager().getDataState().getUser().getLinks().getSelf().getHref());
+			if (environmentURL.getProtocol() == storeUserURL.getProtocol())
 			{
-				startNewActivity(bundle, SignupActivity.class);
+				moveToYonaActivity();
 			}
-		});
-		findViewById(R.id.login).setOnClickListener(new View.OnClickListener()
+			reloadUserFromServer(storeUserURL.toString().replace(storeUserURL.getProtocol(), environmentURL.getProtocol()));
+		}
+		catch (MalformedURLException e)
 		{
-			@Override
-			public void onClick(View v)
-			{
-				startNewActivity(LoginActivity.class);
-			}
-		});
-		findViewById(R.id.environmentSwitch).setOnLongClickListener(new View.OnLongClickListener()
+			AppUtils.reportException(this.getClass().getName(), e, Thread.currentThread());
+		}
+	}
+
+	private void reloadUserFromServer(String url)
+	{
+		if (!AppUtils.isNetworkAvailable(YonaApplication.getAppContext()))
 		{
-			@Override
-			public boolean onLongClick(View v)
-			{
-				switchEnvironment();
-				return true;
-			}
+			AppUtils.displayErrorAlert(this, new ErrorMessage(this.getString(R.string.mandatory_user_fetch_failure)));
+			return;
```

Comment:
The app should exit here. Otherwise, the user can go through the sign-up sequence, thus loosing their account.